crates/host/src/guest.rs

```diff
@@ -66,40 +107,64 @@ pub fn write_slice(ctx: &mut Ctx, guest_ptr: RemotePtr, slice: &[u8]) {
 ///
 /// a better approach is to use an immutable deref from a WasmPtr, which checks against memory
 /// bounds for the guest, and map over the whole thing to a Vec<u8>
-pub fn read_slice(ctx: &mut Ctx, guest_ptr: RemotePtr, len: Len) -> Vec<u8> {
+///
+/// this does the inverse of write_bytes to read a vector of arbitrary length given only a single
+/// GuestPtr value
+///
+/// it reads the first 4 u8 bytes at the GuestPtr position and interprets them as a single u32
+/// value representing a Len which is the length of the return Vec<u8> to read at position
+/// GuestPtr + 4
+///
+/// using the example in write_bytes(), if we had written
+///
+/// ```ignore
+/// [ 3_u8, 0_u8, 0_u8, 0_u8, 1_u8, 2_u8, 3_u8 ]
+/// ```
+///
+/// and this returned a GuestPtr to `5678` then we would read it back by taking the first 4 bytes
+/// at `5678` which would be `[ 3_u8, 0_u8, 0_u8, 0_u8 ]` which we interpret as the length `3_u32`.
+///
+/// we then read the length 3 bytes from position `5682` (ptr + 4) to get our originally written
+/// bytes of `[ 1_u8, 2_u8, 3_u8 ]`.
+pub fn read_bytes(ctx: &Ctx, guest_ptr: GuestPtr) -> Result<Vec<u8>, WasmError> {
     let ptr: WasmPtr<u8, Array> = WasmPtr::new(guest_ptr as _);
-    ptr.deref(ctx.memory(0), 0, len as _)
-        .expect("pointer in bounds")
+    let len_bytes: Vec<u8> = ptr
+        .deref(ctx.memory(0), 0, std::mem::size_of::<Len>() as Len)
+        .ok_or(WasmError::Memory)?
+        .iter()
+        .map(|cell| cell.get())
+        .collect::<Vec<u8>>();
+
+    let len: Len = u32::from_le_bytes([len_bytes[0], len_bytes[1], len_bytes[2], len_bytes[3]]);
```

Comment:
Here's another location where the 4 bytes are assumed instead of being abstracted...

You'd need to do a loop with shifts, something like:

```rust
let mut len: Len = 0;
for i in 0..len_bytes.len() {
    len += (len_bytes[i] << 2 ** i);
}
```

Reply:
^^ untested - be sure do do the math right : )